### PR TITLE
add more secure compilation flags

### PR DESCRIPTION
**Pull request by bratpiorka**
### Description

Add more secure compilation flags + fix errors / warnings.

### Checklist
- [ ] Code compiles without errors locally
- [ ] All tests pass locally
- [ ] CI workflows execute properly

### CODE REVIEW

File: cmake/helpers.cmake [hunk 1]
```diff
@@ -91,18 +91,24 @@ function(add_umf_target_compile_options name)
...
+                    # -flto -fvisibility=hidden
                     $<$<CXX_COMPILER_ID:GNU>:-fdiagnostics-color=auto>)
         if(CMAKE_BUILD_TYPE STREQUAL "Release")
             target_compile_definitions(${name} PRIVATE -D_FORTIFY_SOURCE=2)
         endif()
         if(UMF_DEVELOPER_MODE)
-            target_compile_options(
-                ${name} PRIVATE -Werror -fno-omit-frame-pointer
-                                -fstack-protector-strong)
+            target_compile_options(${name} PRIVATE -fno-omit-frame-pointer)
```
Comment (PatKamin) [inline cmake/helpers.cmake:111]: Change the description of the UMF_DEVELOPER_MODE option as it doesn't enable Werror anymore.

File: cmake/helpers.cmake [hunk 2]
```diff
@@ -91,18 +91,24 @@ function(add_umf_target_compile_options name)
...
+                    # -fPIE
                     -Wall
+                    -Wextra
+                    -Werror
                     -Wpedantic
                     -Wempty-body
                     -Wunused-parameter
+                    -Wformat
+                    -Wformat-security
+                    -fstack-protector-strong
```
Comment (PatKamin) [inline cmake/helpers.cmake:103]: I think this one might make performance suffer as it adds additional checks on a program run. Should it be included only in the Debug builds or left in the DEVELOPER_MODE flag? Same with the FORTIFY_SOURCE.

File: src/ipc_internal.h [hunk 3]
```diff
@@ -12,6 +12,13 @@
 
 #include <umf/base.h>
 
+// disable 4200 warning: nonstandard extension used: zero-sized array in
+// struct/union used in umf_ipc_data_t
+#if defined(_MSC_VER)
+#pragma warning(push)
+#pragma warning(disable : 4200)
```
Comment (lplewa) [inline src/ipc_internal.h:19]: please put it in cmake

File: include/umf/proxy_lib_new_delete.h [hunk 4]
```diff
@@ -48,6 +48,16 @@ SOFTWARE.
 #include <stdlib.h>
 #endif // _WIN32
 
+// disable warning C28196: The requirement that '(_Param_(1)>0)?(return!=0):(1)'
+// is not satisfied. (The expression does not evaluate to true.)
+// disable warning C6387: 'return' could be '0': this does not adhere to the
+// specification for the function 'new'.
```
Comment (lplewa) [inline include/umf/proxy_lib_new_delete.h:54]: well this is a bug - we should check if malloc return null and throw exception in thiscase

File: src/ipc_internal.h (hunk 3 above)
Comment (bratpiorka) [reply src/ipc_internal.h]: done

File: cmake/helpers.cmake (hunk 1 above)
Comment (bratpiorka) [reply cmake/helpers.cmake]: done

File: cmake/helpers.cmake (hunk 2 above)
Comment (bratpiorka) [reply cmake/helpers.cmake]: moved

File: include/umf/proxy_lib_new_delete.h (hunk 4 above)
Comment (bratpiorka) [reply include/umf/proxy_lib_new_delete.h]: fixed

File: src/topology.c [hunk 5]
```diff
@@ -7,9 +7,8 @@
  *
  */
 
-#include <hwloc.h>
-
 #include "base_alloc_global.h"
+#include "provider_os_memory_internal.h"
```
Comment (lplewa) [inline src/topology.c:11]: please move do not create fake dependencies between "objects" - this file do not need os_memory_provider. it needs hwloc

File: test/memspaces/memspace_fixtures.hpp [hunk 6]
```diff
@@ -4,17 +4,17 @@
...
-#include <hwloc.h>
 #include <numa.h>
 #include <numaif.h>
 #include <thread>
+
 #include <umf/memspace.h>
 
+#include "base.hpp"
+#include "memspace_helpers.hpp"
+#include "provider_os_memory_internal.h"
```
Comment (lplewa) [inline test/memspaces/memspace_fixtures.hpp:15]: dito

File: cmake/helpers.cmake [hunk 7]
```diff
@@ -113,12 +116,23 @@ function(add_umf_target_compile_options name)
...
+                    /GS
+                    /analyze
+                    # disable warning 6326: Potential comparison of a constant
+                    # with another constant
+                    /wd6326
+                    # disable 4200 warning: nonstandard extension used:
+                    # zero-sized array in struct/union
+                    /wd4200
+                    /DYNAMICBASE)
+        if(CMAKE_BUILD_TYPE STREQUAL "Release")
```
Comment (PatKamin) [inline cmake/helpers.cmake:134]: You can't rely on this variable during Windows builds. Probably `target_compile_options` supports this construct:
```
$<$<CONFIG:Release>:/sdl>
$<$<CONFIG:Release>:/LTCG>
$<$<CONFIG:Release>:/NXCOMPAT>
```

File: src/base_alloc/base_alloc.c [hunk 8]
```diff
@@ -219,6 +219,12 @@ void *umf_ba_alloc(umf_ba_pool_t *pool) {
     // we'll mark the memory as undefined
     utils_annotate_memory_defined(chunk, sizeof(*chunk));
 
+    // check if the free list is not empty
+    if (pool->metadata.free_list == NULL) {
+        LOG_ERR("pool->metadata.free_list == NULL");
```
[the remark this answers is not in the document]
Comment (bratpiorka) [reply src/base_alloc/base_alloc.c]: done

File: src/cpp_helpers.hpp [hunk 9]
```diff
@@ -145,6 +146,7 @@ umf_memory_provider_ops_t providerMakeCOps() {
         }
 
         if constexpr (std::is_same_v<ParamType, void>) {
+            (void)params; // unused
```
[the remark this answers is not in the document]
Comment (bratpiorka) [reply src/cpp_helpers.hpp]: done

File: src/libumf_windows.c [hunk 10]
```diff
@@ -32,6 +35,10 @@ INIT_ONCE init_once_flag = INIT_ONCE_STATIC_INIT;
 
 BOOL CALLBACK initOnceCb(PINIT_ONCE InitOnce, PVOID Parameter,
                          PVOID *lpContext) {
+    (void)InitOnce;
+    (void)Parameter;
```
[the remark this answers is not in the document]
Comment (bratpiorka) [reply src/libumf_windows.c]: done

File: src/cpp_helpers.hpp [hunk 11]
```diff
@@ -117,6 +117,7 @@ template <typename T, typename ParamType> umf_memory_pool_ops_t poolMakeCOps() {
...
```
[the remark this answers is not in the document]
Comment (bratpiorka) [reply src/cpp_helpers.hpp]: done

File: src/topology.c (hunk 5 above)
Comment (bratpiorka) [reply src/topology.c]: done

File: test/memspaces/memspace_fixtures.hpp (hunk 6 above)
Comment (bratpiorka) [reply test/memspaces/memspace_fixtures.hpp]: done

File: cmake/helpers.cmake (hunk 7 above)
Comment (bratpiorka) [reply cmake/helpers.cmake]: done

**Review by PatKamin — APPROVED** [review]

File: src/provider/provider_os_memory_windows.c [hunk 12]
```diff
@@ -122,7 +122,19 @@ int os_purge(void *addr, size_t length, int advice) {
     // If VirtualFree() succeeds, the return value is nonzero.
     // If VirtualFree() fails, the return value is 0 (zero).
     (void)advice; // unused
+
+    // temporary disable the C6250 warning as we intentionally use the
```
Comment (lukaszstolarczuk) [inline src/provider/provider_os_memory_windows.c:126]: `temporarily` (misspell)

File: src/provider/provider_os_memory_internal.h [hunk 13]
```diff
@@ -8,7 +8,18 @@
 #ifndef UMF_OS_MEMORY_PROVIDER_INTERNAL_H
 #define UMF_OS_MEMORY_PROVIDER_INTERNAL_H
 
+// disable 4100 warning thrown in hwloc.h
+#if defined(_MSC_VER)
+#pragma warning(push)
+#pragma warning(disable : 4100)
+#endif // _MSC_VER
+
 #include <hwloc.h>
```
Comment (lukaszstolarczuk) [inline src/provider/provider_os_memory_internal.h:17]: not `umf_hwloc.h`?

File: CMakeLists.txt [hunk 14]
```diff
@@ -37,8 +37,7 @@ option(UMF_BUILD_BENCHMARKS "Build UMF benchmarks" OFF)
 option(UMF_BUILD_BENCHMARKS_MT "Build UMF multithreaded benchmarks" OFF)
 option(UMF_BUILD_EXAMPLES "Build UMF examples" ON)
 option(UMF_BUILD_GPU_EXAMPLES "Build UMF GPU examples" OFF)
-option(UMF_DEVELOPER_MODE "Enable developer checks, treats warnings as errors"
-       OFF)
+option(UMF_DEVELOPER_MODE "Enable additional developer checks" OFF)
```
Comment (lukaszstolarczuk) [inline CMakeLists.txt:40]: pls update Readme file, as well

File: src/provider/provider_os_memory_internal.h (hunk 13 above)
Comment (bratpiorka) [reply src/provider/provider_os_memory_internal.h:17]: thanks - done

File: CMakeLists.txt (hunk 14 above)
Comment (bratpiorka) [reply CMakeLists.txt:40]: done

File: benchmark/ubench.c [hunk 15]
```diff
@@ -28,6 +28,17 @@
 #include <unistd.h>
 #endif
 
+// disable warning 6308:'realloc' might return null pointer: assigning null
```
[the remark this answers is not in the document]
Comment (bratpiorka) [reply benchmark/ubench.c:41]: partly done - I added a few more comments, but there are still a few places (like ubench.h and hwloc.h) where I added information that I just don't want to change these files

[the remark this answers is not in the document]
Comment (lukaszstolarczuk) [reply benchmark/ubench.c:41]: thx!

**Review by lukaszstolarczuk — APPROVED** [review]

**Pull request merged**